cmd/insights/contributors.go

```diff
@@ -0,0 +1,289 @@
+	return &ErrRepoNotFoundOrIndexed{
+		RepoURL: repoURL,
+	}
+}
+
+func findRepositoryByOwnerAndRepoName(ctx context.Context, apiClient *client.APIClient, repoURL string) (*client.DbRepo, error) {
+	owner, repoName, err := repoquery.GetOwnerAndRepoFromURL(repoURL)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract owner and repo from url: %w", err)
+	}
+	repo, response, err := apiClient.RepositoryServiceAPI.FindOneByOwnerAndRepo(ctx, owner, repoName).Execute()
+	if err != nil {
+		if response.StatusCode == http.StatusNotFound {
+			return nil, NewErrRepoNotFoundOrIndexed(repoURL)
+		}
+		return nil, err
+	}
+	return repo, nil
+}
+
+func findAllRepositoryContributors(ctx context.Context, opts *Options, repoURL string) (*repositoryContributorsInsights, error) {
+	repo, err := findRepositoryByOwnerAndRepoName(ctx, opts.APIClient, repoURL)
+	if err != nil {
+		var errRepoNotFoundOrIndexed *ErrRepoNotFoundOrIndexed
+		if errors.As(err, &errRepoNotFoundOrIndexed) {
+			fmt.Println("ignoring repository issue:", errRepoNotFoundOrIndexed)
+			return nil, nil
+		}
+		return nil, fmt.Errorf("could not get contributor insights for repository %s: %w", repoURL, err)
+	}
+	repoContributorInsights := &repositoryContributorsInsights{
+		RepoID:  repo.Id,
+		RepoURL: repoURL,
+	}
+	var (
+		errorChan = make(chan error)
+		doneChan  = make(chan bool)
+		waitGroup = new(sync.WaitGroup)
+	)
+	waitGroup.Add(4)
+	go func() {
+		defer waitGroup.Done()
+		response, err := findNewRepositoryContributors(ctx, opts.APIClient, repo.Id, opts.Period)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		for _, data := range response.Data {
+			repoContributorInsights.New = append(repoContributorInsights.New, data.AuthorLogin)
+		}
+	}()
+	go func() {
+		defer waitGroup.Done()
+		response, err := findRecentRepositoryContributors(ctx, opts.APIClient, repo.Id, opts.Period)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		for _, data := range response.Data {
+			repoContributorInsights.Recent = append(repoContributorInsights.Recent, data.AuthorLogin)
+		}
+	}()
+	go func() {
+		defer waitGroup.Done()
+		response, err := findChurnRepositoryContributors(ctx, opts.APIClient, repo.Id, opts.Period)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		for _, data := range response.Data {
+			repoContributorInsights.Churn = append(repoContributorInsights.Churn, data.AuthorLogin)
+		}
+	}()
+	go func() {
+		defer waitGroup.Done()
+		response, err := findRepeatRepositoryContributors(ctx, opts.APIClient, repo.Id, opts.Period)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		for _, data := range response.Data {
+			repoContributorInsights.Repeat = append(repoContributorInsights.Repeat, data.AuthorLogin)
+		}
+	}()
+	go func() {
+		waitGroup.Wait()
+		close(doneChan)
+	}()
+	select {
+	case <-doneChan:
+		return repoContributorInsights, nil
+	case err = <-errorChan:
+		close(errorChan)
+		return nil, err
```

Comment:
Same here: only the first error gets popped off the error channel. The others get squashed.